lib/matplotlib/widgets.py

```diff
@@ -1417,6 +1417,543 @@ def _update(self):
+            self.set_midline_marker()
+
+        if self.end_c_lock is True:
+            # If marker a is locked only move end c.
+            pos_a = self.x0, event.xdata
+            pos_b = self.y0, event.ydata
+            self.marker_c.set_data(event.xdata, event.ydata)
+            self.ruler.set_data(pos_a, pos_b)
+            self.set_midline_marker()
+
+        if self.end_b_lock is True:
+            # If marker b is locked shift the whole ruler.
+            b_dx = event.xdata - self.old_mid_coords[0]
+            b_dy = event.ydata - self.old_mid_coords[1]
+            pos_a = self.x0 + b_dx, self.x1 + b_dx
+            pos_b = self.y0 + b_dy, self.y1 + b_dy
+
+            marker_a_coords = self.old_marker_a_coords[0][0] + b_dx, \
+                              self.old_marker_a_coords[0][1] + b_dy
+            marker_c_coords = self.old_marker_c_coords[0][0] + b_dx, \
+                              self.old_marker_c_coords[0][1] + b_dy
+
+            self.ruler.set_data(pos_a, pos_b)
+            self.marker_a.set_data(marker_a_coords)
+            self.marker_b.set_data(event.xdata, event.ydata)
+            self.marker_c.set_data(marker_c_coords)
+
+        self.update_text()
+        self._update_artists()
+
+    def set_midline_marker(self):
+        self.marker_b.set_visible(True)
+        self.marker_b.set_data(self.midline_coords)
+
+    @property
+    def midline_coords(self):
+        pos0, pos1 = self.line_coords
+        mid_line_coords = (pos0[0] + pos1[0]) / 2, (pos0[1] + pos1[1]) / 2
+
+        return mid_line_coords
+
+    def draw_ruler(self, event):
+        """
+        If the left mouse button is pressed and held draw the ruler as the 
+        mouse is dragged
+        """
+
+        self.x1 = event.xdata
+        self.y1 = event.ydata
+
+        # If shift is pressed ruler is constrained to horizontal axis
+        if self.shift_pressed is True:
+            pos_a = self.x0, self.x1
+            pos_b = self.y0, self.y0
+        # If control is pressed ruler is constrained to vertical axis
+        elif self.control_pressed is True:
+            pos_a = self.x0, self.x0
+            pos_b = self.y0, self.y1
+        # Else the ruler follow the mouse cursor
+        else:
+            pos_a = self.x0, self.x1
+            pos_b = self.y0, self.y1
+
+        self.ruler.set_data([pos_a], [pos_b])
+        x1 = self.ruler.get_path().vertices[1][0]
+        y1 = self.ruler.get_path().vertices[1][1]
+
+        self.marker_c.set_visible(True)
+        self.marker_c.set_data(x1, y1)
+
+        self.set_midline_marker()
+        self.update_text()
+        self._update_artists()
+
+    @property
+    def line_coords(self):
+        line_coords = self.ruler.get_path().vertices
+        x0 = line_coords[0][0]
+        y0 = line_coords[0][1]
+        x1 = line_coords[1][0]
+        y1 = line_coords[1][1]
+
+        pos_a = x0, y0
+        pos_b = x1, y1
+
+        return pos_a, pos_b
```

Comment:
Simply because I forgot about `get_data()` . Once again this cleans it up. Thanks.